homeassistant/components/vilfo/config_flow.py

```diff
@@ -0,0 +1,121 @@
+"""Config flow for Vilfo Router integration."""
+import ipaddress
+import logging
+import re
+
+from vilfo import Client as VilfoClient
+from vilfo.exceptions import (
+    AuthenticationException as VilfoAuthenticationException,
+    VilfoException,
+)
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_HOST, CONF_ID, CONF_MAC
+
+from .const import ATTR_DEFAULT_HOST, DOMAIN  # pylint:disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default=ATTR_DEFAULT_HOST): str,
+        vol.Required(CONF_ACCESS_TOKEN, default=""): str,
+    }
+)
+
+
+def host_valid(host):
+    """Return True if hostname or IP address is valid."""
+    try:
+        if ipaddress.ip_address(host).version == (4 or 6):
+            return True
+    except ValueError:
+        disallowed = re.compile(r"[^a-zA-Z\d\-]")
+        return all(x and not disallowed.search(x) for x in host.split("."))
+
+
+async def validate_input(hass: core.HomeAssistant, data):
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    # Validate the host before doing anything else.
+    if not host_valid(data[CONF_HOST]):
+        raise InvalidHost
+
+    config = {}
+
+    # Attempt to connect and call the ping endpoint.
+    # This doesn't validate authentication.
+    controller = VilfoClient(host=data[CONF_HOST], token=data[CONF_ACCESS_TOKEN])
+    try:
+        controller.ping()
```

Comment:
> Is this async safe? Is there blocking I/O here?

Ah, yeah. I think there might be 🤦‍♂ 

I've refactored that so that both `controller.ping()` and `controller.get_board_information()` are called in an executor job.

I will finalize an update to the PR tomorrow 🙂